jax/_src/scipy/signal.py

```diff
@@ -146,3 +151,343 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+
+  if jnp.result_type(win, np.complex64) != outdtype:
+    win = win.astype(outdtype)
+
+  # Determine scale
+  if scaling == 'density':
+    scale = 1.0 / (fs * (win*win).sum())
+  elif scaling == 'spectrum':
+    scale = 1.0 / win.sum()**2
+  else:
+    raise ValueError(f'Unknown scaling: {scaling}')
+  if mode == 'stft':
+    scale = jnp.sqrt(scale)
+
+  # Determine onesided/ two-sided
+  if return_onesided:
+    sides = 'onesided'
+    if jnp.iscomplexobj(x) or jnp.iscomplexobj(y):
+      sides = 'twosided'
+      warnings.warn('Input data is complex, switching to '
+                    'return_onesided=False')
+  else:
+    sides = 'twosided'
+
+  if sides == 'twosided':
+    freqs = jax.numpy.fft.fftfreq(nfft, 1/fs)
+  elif sides == 'onesided':
+    freqs = jax.numpy.fft.rfftfreq(nfft, 1/fs)
+
+  # Perform the windowed FFTs
+  result = _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides)
+
+  if y is not None:
+    # All the same operations on the y data
+    result_y = _fft_helper(y, win, detrend_func, nperseg, noverlap, nfft,
+                           sides)
+    result = jnp.conjugate(result) * result_y
+  elif mode == 'psd':
+    result = jnp.conjugate(result) * result
+
+  result *= scale
+
+  if sides == 'onesided' and mode == 'psd':
+    end = None if nfft % 2 else -1
+    result = result.at[..., 1:end].mul(2)
+
+  time = jnp.arange(nperseg / 2, x.shape[-1] - nperseg / 2 + 1,
+                    nperseg - noverlap) / fs
+  if boundary is not None:
+    time -= (nperseg/2) / fs
+
+  result = result.astype(outdtype)
+
+  # All imaginary parts are zero anyways
+  if y is None and mode != 'stft':
+    result = result.real
+
+  # Output is going to have new last axis for time/window index, so a
+  # negative axis index shifts down one
+  if axis < 0:
+    axis -= 1
+
+  # Move frequency axis back to axis where the data came from
+  result = jnp.moveaxis(result, -1, axis)
+
+  return freqs, time, result
+
+
+@_wraps(osp_signal.stft)
+def stft(x, fs=1.0, window='hann', nperseg=256, noverlap=None, nfft=None,
+         detrend=False, return_onesided=True, boundary='zeros', padded=True,
+         axis=-1):
+  freqs, time, Zxx = _spectral_helper(x, None, fs, window, nperseg, noverlap,
+                                      nfft, detrend, return_onesided,
+                                      scaling='spectrum', axis=axis,
+                                      mode='stft', boundary=boundary,
+                                      padded=padded)
+
+  return freqs, time, Zxx
+
+
+_csd_description = """
+The original SciPy function exhibits slightly different behavior between
+``csd(x, x)``` and ```csd(x, x.copy())```.  The LAX-backend version is designed
+to follow the latter behavior.  For using the former behavior, call this
+function as `csd(x, None)`."""
+
+
+@_wraps(osp_signal.csd, lax_description=_csd_description)
+def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
```

Comment:
Currently, calling `cad(x, y)` without specifying `nperseg` gives very obscure error message.

Reply:
Found that `_triage_segments` must occur first. Reordered and streamlined input verification codes.